src/hb-ot-os2-table.hh

```diff
@@ -30,24 +30,91 @@
 #include "hb-open-type.hh"
 #include "hb-ot-os2-unicode-ranges.hh"
 
-namespace OT {
+#include "hb-set.hh"
 
 /*
  * OS/2 and Windows Metrics
  * https://docs.microsoft.com/en-us/typography/opentype/spec/os2
  */
 #define HB_OT_TAG_OS2 HB_TAG('O','S','/','2')
 
+
+namespace OT {
+
+struct OS2V1Tail
+{
+  inline bool has_data () const { return this != &Null (OS2V1Tail); }
+
+  inline bool sanitize (hb_sanitize_context_t *c) const
+  {
+    TRACE_SANITIZE (this);
+    return_trace (c->check_struct (this));
+  }
+
+  public:
+  HBUINT32	ulCodePageRange1;
+  HBUINT32	ulCodePageRange2;
+  public:
+  DEFINE_SIZE_STATIC (8);
+};
+
+struct OS2V2Tail
+{
+  inline bool has_data () const { return this != &Null (OS2V2Tail); }
+
+  inline bool sanitize (hb_sanitize_context_t *c) const
+  {
+    TRACE_SANITIZE (this);
+    return_trace (c->check_struct (this));
+  }
+
+  public:
+  HBINT16	sxHeight;
+  HBINT16	sCapHeight;
+  HBUINT16	usDefaultChar;
+  HBUINT16	usBreakChar;
+  HBUINT16	usMaxContext;
+  public:
+  DEFINE_SIZE_STATIC (10);
+};
+
+struct OS2V5Tail
+{
+  inline bool has_data () const { return this != &Null (OS2V5Tail); }
+
+  inline bool sanitize (hb_sanitize_context_t *c) const
+  {
+    TRACE_SANITIZE (this);
+    return_trace (c->check_struct (this));
+  }
+
+  public:
+  HBUINT16	usLowerOpticalPointSize;
+  HBUINT16	usUpperOpticalPointSize;
+  public:
+  DEFINE_SIZE_STATIC (4);
+};
+
 struct OS2
 {
   static const hb_tag_t tableTag = HB_OT_TAG_OS2;
 
+  inline bool has_data () const { return this != &Null (OS2); }
+
   inline bool sanitize (hb_sanitize_context_t *c) const
   {
     TRACE_SANITIZE (this);
-    return_trace (c->check_struct (this));
+    if (unlikely (!c->check_struct (this))) return_trace (false);
+    if (unlikely (version >= 5 && !v5Tail.sanitize (c))) return_trace (false);
+    if (unlikely (version >= 2 && !v2Tail.sanitize (c))) return_trace (false);
+    if (unlikely (version >= 1 && !v1Tail.sanitize (c))) return_trace (false);
+    return_trace (true);
   }
 
+  inline const OS2V5Tail *get_v5 () const { return &(version >= 5 ? v5Tail : Null (OS2V5Tail)); }
+  inline const OS2V2Tail *get_v2 () const { return &(version >= 2 ? v2Tail : Null (OS2V2Tail)); }
+  inline const OS2V1Tail *get_v1 () const { return &(version >= 1 ? v1Tail : Null (OS2V1Tail)); }
```

Comment:
Normally we make these return references.  You are returning pointers because you need it to be ;).  Let's try to figure out something else.

Reply:
Added `thiz` into the tails.